File: quant1x/data/kline.py
# -*- coding: UTF-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

import os
import pandas as pd
from dataclasses import dataclass, field
from typing import List, Optional, Any
import logging

from quant1x.exchange import Timestamp
from quant1x.level1 import protocol
from quant1x.level1.client import get_std_conn
from quant1x.level1.security_bars import SecurityBarsRequest, SecurityBarsResponse, KLineType, SecurityBar, SECURITY_BARS_MAX
# from quant1x.factors import base as factors  # 延迟导入以避免循环导入
from quant1x.level1.xdxr_info import XdxrInfo
import quant1x.data.xdxr as xdxr_module
from quant1x.config import config
from quant1x.data import adapter
from quant1x.data.adapter import DataAdapter, PLUGIN_MASK_BASE_DATA, register, DEFAULT_DATA_PROVIDER
from quant1x.data.base import BASE_KLINE

logger = logging.getLogger(__name__)

MAX_KLINE_LOOKBACK_DAYS = 1

@dataclass
class KLine:
    date: str = ""
    open: float = 0.0
    close: float = 0.0
    high: float = 0.0
    low: float = 0.0
    volume: float = 0.0
    amount: float = 0.0
    up: int = 0
    down: int = 0
    datetime: str = ""
    adjustment_count: int = 0

    def adjust(self, adj):
        # 延迟导入以避免循环导入
        from quant1x.factors import base as factors
        adj = factors.CumulativeAdjustment(
            timestamp=adj.timestamp,
            m=adj.m,
            a=adj.a,
            monetary_adjustment=adj.monetary_adjustment,
            share_adjustment_ratio=adj.share_adjustment_ratio,
            no=adj.no
        )
        
        self.open = self.open * adj.m + adj.a
        self.close = self.close * adj.m + adj.a
        self.high = self.high * adj.m + adj.a
        self.low = self.low * adj.m + adj.a
        
        # 成交量复权
        if self.volume != 0:
            # 1. 计算均价
            ap = self.amount / self.volume
            # 2. 均价复权
            ap_adjusted = ap * adj.m + adj.a
            # 3. 成交量复权
            self.volume *= (1 + adj.share_adjustment_ratio)
            # 4. 以新成交量*均价计算成交额
            self.amount = self.volume * ap_adjusted
        
        # 5. 更新除权除息次数
        self.adjustment_count = adj.no

    @staticmethod
    def headers() -> List[str]:
        return ["date", "open", "close", "high", "low", "volume", "amount", "up", "down", "datetime", "adjustment_count"]

def save_kline(filename: str, values: List[KLine]):
    if not values:
        return
        
    dirname = os.path.dirname(filename)
    if dirname:
        os.makedirs(dirname, exist_ok=True)
    
    data = [
        {
            "date": v.date,
            "open": v.open,
            "close": v.close,
            "high": v.high,
            "low": v.low,
            "volume": v.volume,
            "amount": v.amount,
            "up": v.up,
            "down": v.down,
            "datetime": v.datetime,
            "adjustment_count": v.adjustment_count
        }
        for v in values
    ]
    
    df = pd.DataFrame(data, columns=KLine.headers())
    df.to_csv(filename, index=False)

def read_kline_from_csv(filename: str) -> List[KLine]:
    klines = []
    if not os.path.exists(filename):
        return klines
        
    try:
        df = pd.read_csv(filename)
        # Ensure columns exist
        required_cols = KLine.headers()
        if not all(col in df.columns for col in required_cols):
            return klines
            
        for _, row in df.iterrows():
            kline = KLine(
                date=str(row['date']),
                open=float(row['open']),
                close=float(row['close']),
                high=float(row['high']),
                low=float(row['low']),
                volume=float(row['volume']),
                amount=float(row['amount']),
                up=int(row['up']),
                down=int(row['down']),
                datetime=str(row['datetime']),
                adjustment_count=int(row['adjustment_count'])
            )
            klines.append(kline)
    except Exception as e:
        logger.error(f"Failed to read kline csv {filename}: {e}")
        
    return klines

def load_kline(code: str) -> List[KLine]:
    filename = config.get_kline_filename(code)
    logger.debug(f"[dataset::KLine] kline file: {filename}")
    return read_kline_from_csv(filename)

def fetch_kline(code: str, start: int, count: int, kline_type: KLineType = KLineType.DAILY) -> List[SecurityBar]:
    try:
        with get_std_conn() as conn:
            req = SecurityBarsRequest(code, kline_type.value, start, count)
            # Determine is_index from request logic or pass it explicitly if needed
            # SecurityBarsRequest constructor determines is_index
            
            resp = SecurityBarsResponse(req.is_index, kline_type.value)
            protocol.process(conn, req, resp)
            return resp.list
    except Exception as e:
        logger.error(f"[dataset::KLine] fetch_kline error: {e}")
        return []

def apply_forward_adjustment_for_event(klines: List[KLine], 
                                       current_start_date: Timestamp, 
                                       dividends: List[XdxrInfo]):
    if not klines:
        return
        
    # 最后一根K线的日期
    last_day = klines[-1].date
    # 转成时间戳且对齐时间
    ts_last_day = Timestamp.parse(last_day).get_pre_market_time()
    # 计算最后一根K线的下一个交易日的日期
    # TODO: Implement next_trading_day properly. For now, use +1 day approximation or just rely on date comparison
    last_day_next = ts_last_day.offset(hour=24).only_date() # Approximation
    start_date_str = current_start_date.only_date()
    
    # Filter dividends
    xdxr_infos = [
        x for x in dividends 
        if x.Date <= last_day_next and x.Category == 1
    ]
    
    # Sort by date? C++ uses std::views::filter, order depends on input.
    # Assuming dividends are sorted by date.
    
    times = len(xdxr_infos)
    
    for info in xdxr_infos:
        if info.Date <= start_date_str:
            # IPO check logic in C++ is commented out or specific?
            # "除权除息数据在日线第一条数据之前... continue" is commented out in C++.
            pass
        else:
            m, a = info.adjust_factor()
            share_ratio = info.compute_share_adjustment_ratio()
            
            for kline in klines:
                if kline.date >= info.Date:
                    break
                
                if kline.date < info.Date:
                    kline.open = kline.open * m + a
                    kline.close = kline.close * m + a
                    kline.high = kline.high * m + a
                    kline.low = kline.low * m + a
                    
                    if kline.volume != 0:
                        ap = kline.amount / kline.volume
                        ap_adjusted = ap * m + a
                        kline.volume *= (1 + share_ratio)
                        kline.amount = kline.volume * ap_adjusted
                    
                    kline.adjustment_count += 1
        
        times -= 1

class DataKLine(DataAdapter):
    def kind(self) -> int:
        return BASE_KLINE
        
    def owner(self) -> str:
        return DEFAULT_DATA_PROVIDER
        
    def key(self) -> str:
        return "kline"
        
    def name(self) -> str:
        return "前复权K线"
        
    def usage(self) -> str:
        return "前复权K线数据"
        
    def print(self, code: str, dates: Optional[List[Timestamp]] = None) -> None:
        pass
        
    def update(self, code: str, date: Optional[Timestamp] = None) -> None:
        # 1. Determine start date from local cache
        current_start_date = Timestamp.parse("1990-12-19") # market_first_date
        cache_filename = config.get_kline_filename(code)
        cache_klines = read_kline_from_csv(cache_filename)
        
        klines_length = len(cache_klines)
        klines_offset_days = MAX_KLINE_LOOKBACK_DAYS
        adjust_times = 0
        
        if klines_length > 0:
            if klines_offset_days > klines_length:
                klines_offset_days = klines_length
            
            kline = cache_klines[klines_length - klines_offset_days]
            current_start_date = Timestamp.parse(kline.date)
            adjust_times = kline.adjustment_count
            
        # 2. Determine end date
        current_end_date = Timestamp.now().get_pre_market_time()
        logger.debug(f"[dataset::KLine] [{code}]: from {current_start_date.only_date()} to {current_end_date.only_date()}")
        
        step = SECURITY_BARS_MAX
        start = 0
        hs: List[List[SecurityBar]] = []
        element_count = 0
        
        while True:
            count = step
            reply = fetch_kline(code, start, count)
            if not reply:
                break
                
            element_count += len(reply)
            hs.append(reply)
            
            last_bar = reply[-1]
            last_bar_date = Timestamp.parse(f"{last_bar.Year}-{last_bar.Month:02d}-{last_bar.Day:02d}").get_pre_market_time()
            
            if last_bar_date < current_start_date:
                break
                
            if len(reply) < count:
                break
                
            start += count
            
        hs.reverse()
        
        incremental_klines: List[KLine] = []
        
        for vec in hs:
            for row in vec:
                date_time = Timestamp.parse(f"{row.Year}-{row.Month:02d}-{row.Day:02d}").get_pre_market_time()
                
                if date_time < current_start_date or date_time > current_end_date:
                    continue
                    
                kx = KLine(
                    date=date_time.only_date(),
                    open=row.Open,
                    close=row.Close,
                    high=row.High,
                    low=row.Low,
                    volume=row.Vol * 100, # Convert to shares
                    amount=row.Amount,
                    up=row.UpCount,
                    down=row.DownCount,
                    datetime=row.DateTime,
                    adjustment_count=0
                )
                incremental_klines.append(kx)
                
        # 6. Adjustment logic
        is_fresh_fetch_require_adjustment = (adjust_times == 1)
        dividends = xdxr_module.load_xdxr(code)
        
        if is_fresh_fetch_require_adjustment:
            apply_forward_adjustment_for_event(incremental_klines, current_start_date, dividends)
            
        # 7. Merge
        klines = []
        if klines_length > klines_offset_days:
            klines.extend(cache_klines[:klines_length - klines_offset_days])
            
        klines.extend(incremental_klines)
        
        # 8. Forward adjust
        if not is_fresh_fetch_require_adjustment:
            apply_forward_adjustment_for_event(klines, current_start_date, dividends)
            
        # 9. Save
        save_kline(cache_filename, klines)


# 注册插件
_data_kline_plugin = adapter.register(DataKLine)

File: quant1x/data/kline.rs
use crate::data::meta::instrument::Instrument;
use crate::contrib::data::tdx::level1::std::KLineType;
use crate::meta::Timestamp;
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use crate::data::adapter::DataAdapter;
use crate::apply_forward_adjustment_for_event;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct KLine {
    #[serde(rename = "date")]
    pub date: String,
    #[serde(rename = "open")]
    pub open: f64,
    #[serde(rename = "close")]
    pub close: f64,
    #[serde(rename = "high")]
    pub high: f64,
    #[serde(rename = "low")]
    pub low: f64,
    #[serde(rename = "volume")]
    pub volume: f64,
    #[serde(rename = "amount")]
    pub amount: f64,
    #[serde(rename = "up")]
    pub up: i32,
    #[serde(rename = "down")]
    pub down: i32,
    #[serde(rename = "datetime")]
    pub datetime: String,
    #[serde(rename = "adjustment_count")]
    pub adjustment_count: i32,
}

impl KLine {
    pub fn headers() -> Vec<String> {
        vec![
            "date".into(),
            "open".into(),
            "close".into(),
            "high".into(),
            "low".into(),
            "volume".into(),
            "amount".into(),
            "up".into(),
            "down".into(),
            "datetime".into(),
            "adjustment_count".into(),
        ]
    }
}

#[derive(Debug)]
pub struct DataKLine;

impl crate::data::Schema for DataKLine {
    fn kind(&self) -> crate::Kind {
        crate::data::BaseKLine
    }
    fn owner(&self) -> String {
        crate::data::DEFAULT_DATA_PROVIDER.to_string()
    }
    fn key(&self) -> String {
        "day".to_string()
    }
    fn name(&self) -> String {
        "日K线".to_string()
    }
    fn usage(&self) -> String {
        "日K线".to_string()
    }
}

impl crate::data::DataAdapter for DataKLine {
    fn print(&self, _inst: &Instrument, _dates: &[Timestamp]) {}

    fn update(&self, inst: &Instrument, _date: Timestamp) {
        let symbol = inst.symbol();
        // 对齐 Python tdx/kline.py: get_kline_filename(inst, freq)
        //   module_name = freq.cache_key()  # "day"
        //   symbol = inst.symbol()
        //   sub = f"{module_name}/{inst.cache_dir()}"
        //   return f'{config.data_path}/{sub}/{symbol}.csv'
        let filename = get_kline_filename(inst);
        if filename.is_empty() {
            log::error!("[DataKLine] cannot build kline filename for {}", symbol);
            return;
        }
        log::debug!("[DataKLine] cache filename: {}", filename);

        // 确保父目录存在
        let path = std::path::Path::new(&filename);
        if let Some(parent) = path.parent() {
            if let Err(e) = std::fs::create_dir_all(parent) {
                log::error!(
                    "[DataKLine] failed to create parent dir {:?}: {}",
                    parent,
                    e
                );
                return;
            }
        }
        // 遵循 C++ 的逻辑：分页步长（security_bars_max）、回溯天数、合并及除权预处理
        const MAX_KLINE_LOOKBACK_DAYS: usize = 1;
        const SECURITY_BARS_MAX: usize = 800;

        // 先加载已有缓存（若存在），以确定起始日期和调整次数（与 C++ 保持一致）
        let cache_filename = filename.clone();
        let cache_klines: Vec<KLine> = read_kline_from_csv(&cache_filename);
        let klines_length = cache_klines.len();
        let mut klines_offset_days = MAX_KLINE_LOOKBACK_DAYS;
        let mut adjust_times = 0i32;
        // 默认起始日期：1990-12-19（市场首次上市日期）
        let mut current_start_date =
            crate::meta::Timestamp::pre_market_time(1990, 12, 19).unwrap_or(crate::meta::Timestamp::zero());
        if klines_length > 0 {
            if klines_offset_days > klines_length {
                klines_offset_days = klines_length;
            }
            let kline = &cache_klines[klines_length - klines_offset_days];
            // parse date back to Timestamp
            if let Ok(ts) = crate::meta::Timestamp::parse(&kline.date) {
                current_start_date = ts;
            }
            adjust_times = kline.adjustment_count;
        }

        // 从当前时间确定结束日期（尽可能使用今日的盘前时间作为结束日期）
        let current_end_date =
            crate::meta::Timestamp::pre_market_time_from_current(&crate::meta::Timestamp::now())
                .unwrap_or(crate::meta::Timestamp::now());
        // 构建日期范围
        let ts_range = crate::meta::date_range(current_start_date, current_end_date, false);
        if ts_range.is_empty() {
            log::debug!("[DataKLine] empty date range for {}", symbol);
            return;
        }

        let total = ts_range.len();

        // 按日期范围分页从 level1 拉取数据（每页作为一个向量，保持页内顺序，之后再翻转页序以匹配 C++）
        let mut hs: Vec<Vec<crate::contrib::data::tdx::standard::SecurityBar>> = Vec::new();
        let step = SECURITY_BARS_MAX;
        let mut start_idx: usize = 0;
        while start_idx < total {
            let remaining = total - start_idx;
            let count = std::cmp::min(step, remaining) as u16;
            // 注意：start 使用日期索引，datasets 层直接调度 SecurityBars 请求
            match crate::data::kline_raw::fetch_kline(
                &symbol,
                start_idx as u32,
                count,
                KLineType::RiK,
            ) {
                Some(resp) => {
                    let response_count = resp.count as usize;
                    if resp.list.is_empty() {
                        break;
                    }
                    hs.push(resp.list);
                    if response_count < count as usize {
                        break;
                    }
                    start_idx = start_idx.saturating_add(count as usize);
                }
                None => {
                    log::warn!(
                        "[DataKLine] fetch_kline returned None for {} start={}",
                        symbol,
                        start_idx
                    );
                    break;
                }
            }
        }

        // // 如果没有获取到任何页，则不保存CSV（空列表不保存，包括表头）
        // if hs.is_empty() {
        //     return;
        // }
        // C++ 会将分页结果反转为时间升序；保留每页内的顺序，只 reverse 外层 pages
        hs.reverse();

        // 根据抓取到的 pages 并按日期范围筛选，构建增量 K 线列表（按页内顺序迭代）
        let mut incremental_klines: Vec<KLine> = Vec::new();
        for page in hs.iter() {
            for row in page.iter() {
                // 为 bar 日期构造盘前时间戳
                let date_time =
                    crate::meta::Timestamp::pre_market_time(row.year, row.month as u32, row.day as u32)
                        .unwrap_or(crate::meta::Timestamp::now());
                if date_time < ts_range[0] || date_time > ts_range[total - 1] {
                    continue;
                }
                let kx = KLine {
                    date: date_time.only_date(),
                    open: row.open,
                    close: row.close,
                    high: row.high,
                    low: row.low,
                    volume: row.vol * 100.0,
                    amount: row.amount,
                    up: row.up_count as i32,
                    down: row.down_count as i32,
                    datetime: row.datetime.clone(),
                    adjustment_count: 0,
                };
                incremental_klines.push(kx);
            }
        }

        // 判断是否仅针对最新抓取的那一部分需要做预除权调整
        let is_fresh_fetch_require_adjustment = adjust_times == 1;
        // 仅从本地缓存加载除权除息数据（对齐 Python: get_xdxr_list(inst)）
        let dividends = crate::data::xdxr::load_xdxr(&symbol);
        if is_fresh_fetch_require_adjustment {
            apply_forward_adjustment_for_event!(&mut incremental_klines, ts_range[0], &dividends);
        }

        // 按照 C++ 的合并逻辑合并缓存与增量数据
        let mut klines: Vec<KLine> = Vec::new();
        if klines_length > klines_offset_days {
            klines.extend_from_slice(&cache_klines[..(klines_length - klines_offset_days)]);
        }
        if klines.is_empty() {
            klines = incremental_klines.clone();
        } else {
            klines.extend(incremental_klines.into_iter());
        }

        if !is_fresh_fetch_require_adjustment {
            apply_forward_adjustment_for_event!(&mut klines, ts_range[0], &dividends);
        }

        // 持久化保存
        if klines.is_empty() {
            // 空列表不保存CSV，包括表头
            return;
        }
        let tmp = format!("{}.tmp", filename);
        match std::fs::File::create(&tmp) {
            Ok(f) => {
                let mut w = csv::Writer::from_writer(f);
                if let Err(e) = w.write_record(KLine::headers()) {
                    log::error!("[DataKLine] write header failed: {}", e);
                }
                for row in klines.iter() {
                    let rec: Vec<String> = vec![
                        row.date.clone(),
                        row.open.to_string(),
                        row.close.to_string(),
                        row.high.to_string(),
                        row.low.to_string(),
                        row.volume.to_string(),
                        row.amount.to_string(),
                        row.up.to_string(),
                        row.down.to_string(),
                        row.datetime.clone(),
                        row.adjustment_count.to_string(),
                    ];
                    if let Err(e) = w.write_record(rec) {
                        log::error!("[DataKLine] write row failed: {}", e);
                    }
                }
                let _ = w.flush();
                if let Err(e) = std::fs::rename(&tmp, &filename) {
                    log::error!("[DataKLine] rename failed {} -> {}: {}", tmp, filename, e);
                }
            }
            Err(e) => {
                log::error!("[DataKLine] create tmp {} failed: {}", tmp, e);
            }
        }
    }
}

// 辅助函数：从 CSV 读取 K 线，功能类似 C++ 的 read_kline_from_csv
pub fn load_klines(filename: &str) -> Vec<KLine> {
    // 使用基于 Serde 的 CSV 反序列化，使字段名可自动与表头匹配。
    let mut klines: Vec<KLine> = Vec::new();
    match std::fs::File::open(filename) {
        Ok(f) => {
            let mut rdr = csv::ReaderBuilder::new().has_headers(true).from_reader(f);
            // Deserialize all records into Vec<KLine> in one go; CSV+Serde maps headers -> struct fields
            match rdr
                .deserialize::<KLine>()
                .collect::<Result<Vec<KLine>, csv::Error>>()
            {
                Ok(v) => klines = v,
                Err(e) => log::error!(
                    "[DataKLine] failed to deserialize kline file {}: {}",
                    filename,
                    e
                ),
            }
        }
        Err(_) => { /* missing file -> return empty vector */ }
    }
    klines
}

fn read_kline_from_csv(filename: &str) -> Vec<KLine> {
    load_klines(filename)
}

/// 生成 K 线缓存文件路径，与 Python tdx/kline.py 的 get_kline_filename(inst, freq) 对齐
///   module_name = freq.cache_key()  # "day"
///   symbol = inst.symbol()
///   sub = f"{module_name}/{inst.cache_dir()}"
///   return f'{config.data_path}/{sub}/{symbol}.csv'
fn get_kline_filename(inst: &Instrument) -> String {
    let symbol = inst.symbol();
    let sub = format!("day/{}", inst.cache_dir());
    format!("{}/{}/{}.csv", crate::config::default_cache_path(), sub, symbol)
}

pub fn init() {
    let plugin = Arc::new(DataKLine) as Arc<dyn crate::data::DataAdapter>;
    crate::data::register(plugin);
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::data::market::detect_symbol;

    #[test]
    fn test_kline_update() {
        let adapter = DataKLine;
        let code = "sz002350";
        let inst = detect_symbol(code);
        assert!(inst.can_construct_symbol());
        let date = Timestamp::now();
        adapter.update(&inst, date);
    }
}
